sharding/receipt_consuming_tx_utils.py

```diff
@@ -0,0 +1,149 @@
+    '''
+    if state.gas_used + tx.startgas > state.gas_limit:
+        return False
+
+    return True
+
+
+def is_valid_receipt_consuming_tx(mainchain_state, shard_state, shard_id, tx):
+    if (tx.v != 1) or (tx.s != 0) or not isinstance(tx.r, int):
+        return False
+    if not tx.to or tx.to == CREATE_CONTRACT_ADDRESS:
+        return False
+    if not simplified_validate_transaction(shard_state, tx):
+        return False
+    receipt_id = tx.r
+    receipt_shard_id = call_valmgr(mainchain_state, 'get_receipts__shard_id', [receipt_id])
+    receipt_value = call_valmgr(mainchain_state, 'get_receipts__value', [receipt_id])
+    if receipt_value <= 0:
+        return False
+    receipt_to = call_valmgr(mainchain_state, 'get_receipts__to', [receipt_id])
+    if ((receipt_shard_id != shard_id) or
+        (receipt_value != tx.value) or
+        (receipt_to != hex(utils.big_endian_to_int((tx.to)))) or
+        call_urs(shard_state, shard_id, 'get_used_receipts', [receipt_id])):
+        return False
+    return True
+
+
+def send_msg_add_used_receipt(state, shard_id, receipt_id):
+    ct = get_urs_ct(shard_id)
+    urs_addr = get_urs_contract(shard_id)['addr']
+    return call_contract_inconstantly(
+        state, ct, urs_addr, 'add_used_receipt', [receipt_id],
+        0, sender_addr=urs_addr
+    )
+
+
+def send_msg_transfer_value(mainchain_state, shard_state, shard_id, tx):
+    urs_addr = get_urs_contract(shard_id)['addr']
+    log_rctx.debug("Begin: urs.balance={}, tx.to.balance={}".format(shard_state.get_balance(urs_addr), shard_state.get_balance(tx.to)))
+
+    receipt_id = tx.r
+    # we should deduct the startgas of this message in advance, because
+    # the message may be possibly a contract, not only a normal value transfer.
+    value = tx.value - tx.gasprice * tx.startgas
+    log_rctx.debug("value={}, tx.value={}, tx.gasprice={}, tx.startgas={}".format(value, tx.value, tx.gasprice, tx.startgas))
+    if value <= 0:
+        return False, None
+
+    # start transactioning
+    if not send_msg_add_used_receipt(shard_state, shard_id, receipt_id):
+        return False, None
+
+    receipt_sender_hex = call_valmgr(mainchain_state, 'get_receipts__sender', [receipt_id])
+    receipt_data = call_valmgr(mainchain_state, 'get_receipts__data', [receipt_id])
+    data = (b'00' * 12) + utils.parse_as_bin(receipt_sender_hex) + receipt_data
+    msg = vm.Message(urs_addr, tx.to, value, tx.startgas - tx.intrinsic_gas_used, data)
+    env_tx = Transaction(0, tx.gasprice, tx.startgas, b'', 0, b'')
+    env_tx._sender = utils.parse_as_bin(receipt_sender_hex)
+    ext = VMExt(shard_state, env_tx)
+    log_rctx.debug("before apply_msg: urs_addr.balance={}, tx.to.balance={}".format(shard_state.get_balance(urs_addr), shard_state.get_balance(tx.to)))
+    # even if `transfer_value` in `apply_msg` fails, no error occurs.
+    # it seems no raise in apply_msg
+    result, gas_remained, data = apply_msg(ext, msg)
```

Comment:
Since a variable `data` was declared before, how about using another variable name instead of `data` here?

Reply:
Sounds good, to prevent from misusing the variables.